lib/segmentation/chainsaw.py

```diff
@@ -0,0 +1,41 @@
+"""
+Chainsaw predicts domain segmentation based on a deep learning model. It uses
+the `stride` tool to determine secondary structure, and produces a TSV file
+with a description of the output.
+
+Source: <https://github.com/JudeWells/Chainsaw>
+"""
+
+import csv
+from collections.abc import Iterator
+from typing import Tuple
+
+from . import SegmentationParser, ParsingError
+
+
+class Parser(SegmentationParser):
+    def __init__(self, csv_path):
+        super().__init__(csv_path)
+
+    def parse(self) -> Iterator[Tuple[str, int, str]]:
+        (csv_path,) = self.paths
+
+        rows = _read_csv_rows(csv_path, delimiter='\t')
+
+        try:
+            data         = rows[0]
+            domain_count = data['ndom']
+            chopping     = data['chopping']
+        except (IndexError, KeyError) as e:
+            raise ParsingError("[Chainsaw] Invalid input format") from e
```

Comment:
This syntax is something I learned about recently -- the error is thrown, but the "cause" of the error is retained, so the backtrace show the new error and it also shows the original one and its full backtrace.